File: Terraform/lambda_functions/get_timesheets.py
import json
import os
import boto3
from datetime import datetime, timedelta
from decimal import Decimal
from boto3.dynamodb.conditions import Key
from botocore.exceptions import ClientError
from auth_utils import get_cors_headers, handle_cors_preflight, get_user_from_token

def decimal_to_float(obj):
    """Convert Decimal objects to float for JSON serialization"""
    if isinstance(obj, Decimal):
        return float(obj)
    elif isinstance(obj, dict):
        return {k: decimal_to_float(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [decimal_to_float(item) for item in obj]
    else:
        return obj

def lambda_handler(event, context):
    headers = get_cors_headers()
    
    # Handle preflight OPTIONS request
    cors_response = handle_cors_preflight(event)
    if cors_response:
        return cors_response

    try:
        # Get and validate user from token
        user_id, auth_error = get_user_from_token(event)
        if auth_error:
            return auth_error

        # Get query parameters
        query_params = event.get('queryStringParameters') or {}
        start_date = query_params.get('start_date') or query_params.get('startDate')
        end_date = query_params.get('end_date') or query_params.get('endDate')
        period = query_params.get('period')

        dynamodb = boto3.resource('dynamodb')
        time_entries_table = dynamodb.Table(os.environ['TIME_ENTRIES_TABLE'])

        # Build query
        query_kwargs = {
            'KeyConditionExpression': Key('user_id').eq(user_id),
            'ScanIndexForward': False  # Most recent first
        }

        # Add date filtering if provided
        if start_date and end_date:
            query_kwargs['FilterExpression'] = Key('date').between(start_date, end_date)
        elif period:
            now = datetime.utcnow()
            
            if period == 'today':
                filter_date = now.strftime('%Y-%m-%d')
                query_kwargs['FilterExpression'] = Key('date').eq(filter_date)
            elif period == 'this-week':
                start_of_week = now - timedelta(days=now.weekday())
                end_of_week = start_of_week + timedelta(days=6)
                query_kwargs['FilterExpression'] = Key('date').between(
                    start_of_week.strftime('%Y-%m-%d'),
                    end_of_week.strftime('%Y-%m-%d')
                )
            elif period == 'this-month':
                start_of_month = now.replace(day=1)
                # Get last day of month
                if now.month == 12:
                    end_of_month = now.replace(year=now.year + 1, month=1, day=1) - timedelta(days=1)
                else:
                    end_of_month = now.replace(month=now.month + 1, day=1) - timedelta(days=1)
                
                query_kwargs['FilterExpression'] = Key('date').between(
                    start_of_month.strftime('%Y-%m-%d'),
                    end_of_month.strftime('%Y-%m-%d')
                )

        try:
            response = time_entries_table.query(**query_kwargs)
            items = response['Items']
            
            # Group entries by date and calculate daily totals
            grouped_entries = {}
            total_hours = 0

            for item in items:
                date = item['date']
                if date not in grouped_entries:
                    grouped_entries[date] = {
                        'date': date,
                        'entries': [],
                        'totalHours': 0
                    }
                
                grouped_entries[date]['entries'].append(item)
                if item.get('type') == 'clock_out' and item.get('hours'):
                    grouped_entries[date]['totalHours'] += float(item['hours'])
                    total_hours += float(item['hours'])

            # Convert to array and sort by date
            timesheets = list(grouped_entries.values())
            timesheets.sort(key=lambda x: x['date'], reverse=True)

            # Convert all data to be JSON serializable
            response_data = {
                'timesheets': timesheets,
                'entries': [entry for day in timesheets for entry in day['entries']],
                'totalHours': round(total_hours, 2),
                'period': period or 'custom',
                'startDate': start_date,
                'endDate': end_date
            }
            
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps(decimal_to_float(response_data))
            }
            
        except ClientError as e:
            print(f"DynamoDB error: {e}")
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({'error': 'Database error'})
            }

    except Exception as e:
        print(f"Error getting timesheets: {e}")
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': 'Internal server error'})
        }

File: Backend/Terraform/lambda_functions/get_timesheets.js
const { connectToDatabase, COLLECTIONS } = require('./utils/database');
const { authenticateUser, createResponse } = require('./utils/auth');
const { formatTimeEntry } = require('./utils/schemas');
const { validateDate, validatePeriod, sanitizeMongoQuery } = require('./utils/validation');

exports.handler = async (event) => {
  console.log('Get timesheets request:', JSON.stringify(event, null, 2));

  try {
    // Handle CORS preflight
    if (event.httpMethod === 'OPTIONS') {
      return createResponse(200, { message: 'CORS preflight' });
    }

    // Authenticate user
    const user = await authenticateUser(event);
    
    // Parse and validate query parameters
    const queryParams = event.queryStringParameters || {};
    const { period = 'today', startDate, endDate } = queryParams;

    // Validate period
    if (!validatePeriod(period)) {
      return createResponse(400, {
        success: false,
        error: 'Invalid period parameter'
      });
    }

    // Validate dates if provided
    if (startDate && !validateDate(startDate)) {
      return createResponse(400, {
        success: false,
        error: 'Invalid startDate format. Use YYYY-MM-DD'
      });
    }

    if (endDate && !validateDate(endDate)) {
      return createResponse(400, {
        success: false,
        error: 'Invalid endDate format. Use YYYY-MM-DD'
      });
    }

    // Calculate date range based on period
    let dateFilter = {};
    const now = new Date();
    
    switch (period) {
      case 'today':
        const today = now.toISOString().split('T')[0];
        dateFilter = { date: today };
        break;
        
      case 'yesterday':
        const yesterday = new Date(now);
        yesterday.setDate(yesterday.getDate() - 1);
        dateFilter = { date: yesterday.toISOString().split('T')[0] };
        break;
        
      case 'this-week':
        const weekStart = new Date(now);
        weekStart.setDate(now.getDate() - now.getDay());
        dateFilter = { 
          date: { 
            $gte: weekStart.toISOString().split('T')[0],
            $lte: now.toISOString().split('T')[0]
          }
        };
        break;
        
      case 'last-week':
        const lastWeekEnd = new Date(now);
        lastWeekEnd.setDate(now.getDate() - now.getDay() - 1);
        const lastWeekStart = new Date(lastWeekEnd);
        lastWeekStart.setDate(lastWeekEnd.getDate() - 6);
        dateFilter = { 
          date: { 
            $gte: lastWeekStart.toISOString().split('T')[0],
            $lte: lastWeekEnd.toISOString().split('T')[0]
          }
        };
        break;
        
      case 'this-month':
        const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
        dateFilter = { 
          date: { 
            $gte: monthStart.toISOString().split('T')[0],
            $lte: now.toISOString().split('T')[0]
          }
        };
        break;
        
      case 'last-month':
        const lastMonthStart = new Date(now.getFullYear(), now.getMonth() - 1, 1);
        const lastMonthEnd = new Date(now.getFullYear(), now.getMonth(), 0);
        dateFilter = { 
          date: { 
            $gte: lastMonthStart.toISOString().split('T')[0],
            $lte: lastMonthEnd.toISOString().split('T')[0]
          }
        };
        break;
        
      case 'custom':
        if (startDate && endDate) {
          // Additional validation: startDate should be before endDate
          if (new Date(startDate) > new Date(endDate)) {
            return createResponse(400, {
              success: false,
              error: 'startDate must be before endDate'
            });
          }
          
          dateFilter = { 
            date: { 
              $gte: startDate,
              $lte: endDate
            }
          };
        } else {
          dateFilter = { date: now.toISOString().split('T')[0] };
        }
        break;
        
      case 'all':
      default:
        // No date filter - get all entries
        break;
    }

    // Connect to database
    const { db } = await connectToDatabase();
    
    // Query time entries with sanitized filters
    const baseQuery = { user_id: user.userId };
    const sanitizedDateFilter = sanitizeMongoQuery(dateFilter);
    const query = { ...baseQuery, ...sanitizedDateFilter };
    
    const entries = await db.collection(COLLECTIONS.TIME_ENTRIES)
      .find(query)
      .sort({ timestamp: 1 })
      .limit(1000) // Prevent excessive data retrieval
      .toArray();

    // Group entries by date and calculate hours
    const groupedByDate = {};
    let totalHours = 0;

    entries.forEach(entry => {
      const date = entry.date;
      
      if (!groupedByDate[date]) {
        groupedByDate[date] = {
          date,
          entries: [],
          totalHours: 0
        };
      }

      const formattedEntry = formatTimeEntry(entry);
      groupedByDate[date].entries.push(formattedEntry);
      
      if (entry.type === 'clock_out' && entry.hours) {
        groupedByDate[date].totalHours += entry.hours;
        totalHours += entry.hours;
      }
    });

    // Convert to array and sort by date (newest first)
    const timesheets = Object.values(groupedByDate).sort((a, b) => 
      new Date(b.date) - new Date(a.date)
    );

    const responseData = {
      success: true,
      data: {
        timesheets,
        entries: entries.map(formatTimeEntry),
        totalHours: parseFloat(totalHours.toFixed(2)),
        period,
        startDate: startDate || null,
        endDate: endDate || null
      }
    };

    return createResponse(200, responseData);

  } catch (error) {
    console.error('Get timesheets error:', error);
    
    if (error.message === 'Invalid token' || error.message === 'No token provided') {
      return createResponse(401, {
        success: false,
        error: 'Unauthorized'
      });
    }

    return createResponse(500, {
      success: false,
      error: 'Internal server error'
    });
  }
};